### DISCUSSION TITLE
Adapter overrides and escape hatches

### DISCUSSION THREAD

**Discussion by jadbox** [withastro/roadmap#725] [Proposal]
### Body

# Summary

Astro adapter interfaces should allow selective overrides of any esbuild directive

# Background & Motivation

Astro adapters are not flexible enough to handle edge cases, and require frequent maintenance forcing delays from clients using new capabilities enabled by the hosting provider. These problems are presently due to adapters explicitly setting their own directives on the esbuild configurations. This has caused a lot of time lost due to teams getting blocked until adapters are updates, versus patching the esbuild. Below are two examples, but there's many services that are blocked waiting on adapter updates.

- The `node:crypto` package newly added but unsupported due to adapter: https://github.com/withastro/astro/issues/8784
- Several (unused) APIs block PG from compiling by the adapter: https://github.com/withastro/astro/issues/8761
- Blocker using cloudflare:* new packages: https://github.com/withastro/astro/pull/8766
- Many service bindings missing like Cloudflare AI that are active blockers 

# Goals

There should be an "escape hatch" so that end users can add/override esbuild configurations (+ maybe other adapter settings). This is especially useful for Cloudflare as the runtime frequently updates, and each time necessitates some patch on the adapter.

# Example

I'm not sure what the ideal fix here is. As a quick stab at the common blocker issues, the below example shows a new adapter interface for esbuild property changes. In the example below, `node:XYZ` is not already enabled in the adapter and will fail in astro build without an adapter update or using a new esbuild override method like the below:

astro.config.ts
```
adapter: cloudflare({
    esbuild: {
      external: ["cloudflare:sockets", "node:XYZ"]
    }
  })
  ```

Perhaps the Vite config inside of `astro.config.ts` always selectively overrides the adapter esbuild directives (under the adapter layer). This way the adapter developer doesn't need to follow a convention. On the flip side, these overrides may break things but that's the risk of using them.

**Comment by jadbox** [comment]
I realize this may complicate debugging adapters when Astro end users can manually override build properties, but the flip side is having teams reject using Astro because the adapters become gatekeepers to every host change or feature addition.

**Reply by alexanderniebuhr** [reply]
Exactly this is the risk, which has to be considered.

**Comment by lilnasy** [comment]
This is a great idea! Astro has vite, postcss and rollup subconfigs. It makes sense for adapters to offer up similar subconfigs to their specific tools as well.

**Reply by alexanderniebuhr** [reply]
Agree, as long as it stays a seamless experience using the adapter and debugging it for 90% for non edge-cases.

**Comment by jadbox** [comment]
I've submitted a PR for the Astro cloudflare adapter that adds the appropriate escape hatches for future platform updates.
https://github.com/withastro/astro/pull/8787

**Comment by alexanderniebuhr** [comment]
Thank you for this discussion topic!
I'd like to address some of the statements and try to better understand your issue:

> This has caused a lot of time lost due to teams getting blocked until adapters are updates, versus patching the esbuild.

I think this statement is rather misleading. While I understand that it might be annoying to wait for the adapter update. There is always the way to just patch the installed npm package & fix it for you until the new update is published. The work required for that, is equal to the work required to update your overrides.

> There should be an "escape hatch"

What exactly would be the difference to patching the npm package on your side?

> Many service bindings missing like Cloudflare AI that are active blockers

I'm unsure what you mean by this, can you please explain. All service bindings are supported if you use `wrangler` & all of miniflare's bindings are supported in the new runtime option.

**Comment by alexanderniebuhr** [comment]
It looks like the svelte team finally has decided in favor of giving users the override options, even if they had some arguments & push-backs in the beginning. I still think those are valid, but giving the user more flexibility seems a good think, as long as we can make sure the default stays as seamless as before. **Which way do we want to follow?**

P.S.: I also saw in one comment, that it Rollup should be able to not chunk, which would allow to get rid of the esbuild step all together.

references:
https://github.com/sveltejs/kit/pull/10424
https://github.com/sveltejs/kit/pull/10521
https://github.com/sveltejs/kit/pull/9398
https://github.com/sveltejs/kit/issues/3733
https://github.com/sveltejs/kit/pull/1914
https://github.com/sveltejs/kit/pull/10544
https://github.com/sveltejs/kit/issues/10028

**Reply by alexanderniebuhr** [reply]
> I fear that people will start opening issues in our repo, when a possible override issue is caused by an adapter/integration .How do users/maintainers know that? 
Solve this issue and I'm on board

https://discord.com/channels/830184174198718474/845430950191038464/1161208642624815205

> That is also my personal fear, and I feel we should solve it first. I think this was also the initial thoughts of the Svelte team. The question is do we need to solve it, or can we workaround with docs, to make clear we don't offer support for custom overrides. 

https://discord.com/channels/830184174198718474/845430950191038464/1161214248819642438

> I guess you could should a warning on build or something
Like "Using configured overrides for this build" or something

https://discord.com/channels/830184174198718474/845430950191038464/1161220913036009524

> I strongly feel it's better for the health of the project to allow flexible control over the adapter build process. Otherwise the plugin will need constant patching every time a package and Miniflare changes.
Cloudflare/Wrangler APIs change multiple times a month (sometimes it's removing a pkg). Advanced users will absolutely need better control of the esbuild externals and Miniflare parameters that fits their team/project. There is no general case workaround possible for advanced users.

https://discord.com/channels/830184174198718474/845430950191038464/1161282625332326492

> I feel the same way, although we should not jump into this without a way to properly protect us (the users and the maintainers). Astro has many features, and as downside us maintainers need to spend a lot of time at debugging issues. And this burnt us. That's why we moved our adapters out from our repository.
Tthe problem we need to solve is not "you must give us this feature because power users we need it", but "how can Astro give the feature without flooding our repos of issue, and redirect these issues to integration maintainers".
Alex mentioned a solution for example: good documentation

https://discord.com/channels/830184174198718474/845430950191038464/1161284845473890385

> Just to repeat one of my suggestions for a short-term solution, if you are an advanced user and have the knowledge to set your own parameters, you should be able to use pnpm or patch-package to patch the adapter while waiting for an update of the adapter.
If you see that your parameters can benefit everyone, you can even contribute those things to the adapter as an PR. This way not only you could benefit from it - e.g. Miniflare parameters - , but also the community can benefit. Which in fact is a healthy open source community. 
I think as Emanuele already backed-up, we are not totally against it. We just want to think this trough carefully, before adding the option.
Additionally to good documentation, another solution was proposed by Erika, e.g. adding logging warnings

https://discord.com/channels/830184174198718474/845430950191038464/1161285398203478106

> There's a team here, so package patching wouldn't be a practical option. Alternatively for our team could fork and publish our own public CF adapter. 

https://discord.com/channels/830184174198718474/845430950191038464/1161287000226271293

> Package patching works with teams, not sure what your setup is, and why it wouldn't work.
MR=PR, just different terms.
I think we laid out some options, adding docs & logging to the PR you have open, if you love to see adding the options for advanced users. 
Being active & contribute whenever you see that something is out off sync, if the PRs have everything included (e.g. docs, description, tests). I would say our time to get it merged is on average <48h.
I also would like to stress that the Cloudflare adapter didn't had Miniflare built in for a long time, and somehow it worked out. Now that it is built in, we are happy to get feedback about it, and improve on it. But having the expectation that it should support every edge-case from day 1 seems a little to ambitious

https://discord.com/channels/830184174198718474/845430950191038464/1161287469950582805

**Comment by alexanderniebuhr** [comment]
@jadbox I actually think this might some of the issues you are facing: https://github.com/cloudflare/workers-sdk/issues/4082
